docs/api_docs.rst

```diff
@@ -0,0 +1,311 @@
+
+.. figure:: figures/api-docs-fig-01.png
+
+   Figure 1: Abstract Interface Inheritance.
+
+---------------------------
+Perturbed Image Generation
+---------------------------
+
+The PerturbImage interface abstracts the behavior of taking a reference image and generating some number perturbations of the image along with paired mask matrices indicating where perturbations have occurred and to what amount.
+
+Implementations should impart no side effects upon the input image.
+
+Immediate candidates for implementation of this interface are occlusion-based saliency algorithms [3] that perform perturbations on image pixels.
+
+Method: ``perturb``  (alias: ``__call__``)
+
+This method transforms an input reference image into a number of mask matrices indicating algorithm-determined regions of perturbation. The input image should be in the form of a `numpy.ndarray` matrix with a shape of either [H x W] or [H x W x C].
+
+Output mask matrix should be 3-dimensional with the format [nMasks x Height x Width], sharing the same height and width to the input reference image matrix height and width dimensions. The implementing algorithm may determine the quantity of output masks per input image, as well as the data-type of the matrix (bool, uint8, float32, etc.). These masks should indicate regions in the reference image that the algorithm has determined should be modified for black-box input. Mask values should be in the [0, 1] range, where a value closer to 1.0 indicates areas of the image that are unperturbed. Note that output mask matrices may be of a floating-point type in order to allow for fractional perturbation.
+
+Inputs
+
+* Reference image (numpy.ndarray)
+
+Outputs
+
+* Masks matrix of type ``np.ndarray`` with shape [nMasks x Height x Width].
+
+Interface: PerturbImage
+------------------------
+.. autoclass:: xaitk_saliency.interfaces.perturb_image.PerturbImage
+   :members:
+
+-----------------------------------------
+Image Occlusion via Perturbation Masks
+-----------------------------------------
+
+A common intermediate step in this process is applying the generated perturbation masks to imagery to produce occluded images. We provide two utility functions as baseline implementation to perform this step. One is the version that performs the transformation as a batch operation, ``xaitk_saliency.utils.masking.occlude_image_batch``. The other, ``xaitk_saliency.utils.masking.occlude_image_streaming``, performs the transformation in a streaming method with optional parallelization in that streaming. While the batch version is simpler and in many cases the faster of the two versions, the streaming version may be more applicable to large image masks or when a great deal of masks are being input, where in such cases the batch version would exceed available memory.
```

Comment:
This section will want added to it some use of the `.. autofunction:: ` directive to render docs from those source python functions.